### PR TITLE
Migrate Syscollector to Inventory at new-agent repository

### PR DESCRIPTION
**Pull request by nbertoldo**
This PR migrates Syscollector module from the wazuh/wazuh repository to the wazuh/wazuh-agent repository as described in #17. 

## Description
This PR migrates the Syscollector module from the wazuh/wazuh repository to the wazuh/wazuh-agent repository, following the tasks outlined in [#15](https://github.com/wazuh/wazuh-agent/issues/15). This migration is part of the ongoing effort to modularize components in the agent repository. It involves refactoring code, adapting to the new repository structure, and ensuring that the migrated modules (Syscollector and DBsync) function correctly in the new environment.

### Code Migration:
- Migrated the Syscollector and DBsync code from the wazuh/wazuh repository to the wazuh/wazuh-agent repository.
- Refactored classes, functions, and variable names to fit the new modular architecture (renamed "Syscollector" to "Inventory").
- Updated the CMake files and added new dependencies supported by vcpkg.

### Integration with Queue:
- Integrated Syscollector (now Inventory) with the new Queue component, ensuring the proper handling of messages.

### Dependencies:
- Migrated dependencies such as libdb, pacman, procps, and rpm to vcpkg where possible.
- Removed unsupported or redundant libraries, including rsync.

### Testing & Validation:
- Added new tests for the migrated modules (Inventory, DBsync, and the new module manager).
- Ensured compatibility with Windows by adapting the code for MSVC compilation.
- Refactored logging mechanisms to align with the new Logger implementation.

### Code Cleanup:
- Removed legacy code for unsupported operating systems.
- Refined the code by analyzing if the http-request used by data-provider could be removed.

## Proposal
- Complete migration and ensure all related modules (Inventory, DBsync) work correctly after refactoring and integration.
- Address any missing dependencies by researching how to integrate unsupported packages into vcpkg or finding alternatives.
- Finalize tests and run validation to ensure no regressions occur in functionality.

## Tests
- Compilation succeeded across all supported platforms (including Windows with MSVC).
- Validated package installation and upgrades.
- Unit and integration tests for Syscollector (Inventory) and DBsync passed successfully.

### Tasks

- [x] Identify the code in the wazuh/wazuh repository.
- [x] Migrate the code to the wazuh/wazuh-agent repository.
- [x] Refactor the code as necessary to fit the new repository structure.
- [x] Test the migrated code to ensure it works correctly in the new repository.

### CODE REVIEW

File: src/common/utils/include/shared.h [hunk 1]
```diff
@@ -110,6 +110,28 @@
...
+#include "os_regex.h"
+#include "string_op.h"
+#include "binaries_op.h"
+#include "os_ip.h"
+#include "os_net.h"
+#include "expression.h"
+#include "randombytes.h"
+#include "bzip2_op.h"
+#include "validate_op.h"
+
```
Comment (TomasTurina) [inline src/common/utils/include/shared.h:133]: What about `version_op.h`? Isn't it necessary here?

File: src/modules/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,17 @@
+cmake_minimum_required(VERSION 3.22)
+
+project(ModuleManager)
+
+set(CMAKE_CXX_STANDARD 20)
```
Comment (TomasTurina) [inline src/modules/CMakeLists.txt:5]: Check if you can use `set_common_settings` here (`src/cmake/CommonSettings.cmake`).

File: src/modules/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+set(CMAKE_CXX_STANDARD_REQUIRED True)
+
+add_subdirectory(inventory)
+
+add_library(ModuleManager src/moduleManager.cpp)
```
Comment (TomasTurina) [inline src/modules/CMakeLists.txt:10]: Check also the usage of `configure_target` here (`src/cmake/ConfigureTarget.cmake`).

File: src/agent/src/agent.cpp [hunk 4]
```diff
@@ -28,6 +29,8 @@ Agent::~Agent()
 
 void Agent::Run()
 {
+    Configuration config;
```
Comment (TomasTurina) [inline src/agent/src/agent.cpp:32]: You should use the `ConfigurationParser` to read the modules configuration. Check how we do with the `communicator` here:
https://github.com/wazuh/wazuh-agent/pull/45/files#diff-da3d5e8dd4752020c509b5b337eb7254314aad465439e743138acdb52912a34cR20

File: src/external/libdb/LICENSE
Comment (TomasTurina) [inline src/external/libdb/LICENSE:1]: This folder (`external`) shouldn't exist. External dependencies must be downloaded.

File: src/modules/include/moduleWrapper.hpp [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+#include <string>
+#include "configuration.hpp"
+
+using namespace std;
+
+struct ModuleWrapper {
+    function<void()> start;
+    function<int(const Configuration&)> setup;
+    function<void()> stop;
+    function<string(const string&)> command;
```
Comment (TomasTurina) [inline src/modules/include/moduleWrapper.hpp:13]: Just a question, is this the function to execute a command?
It will also need a list of parameters and a callback, but we can modify this later.

File: src/modules/inventory/CMakeLists.txt [hunk 6]
```diff
@@ -1,208 +1,59 @@
...
+    ${COMMON_FOLDER}/string_op/include
+    ${COMMON_FOLDER}/time_op/include
+    ${COMMON_FOLDER}/timeHelper/include
+    ${COMMON_FOLDER}/utils)
+
+target_link_directories(inventory PUBLIC ${SRC_FOLDER}/external/rpm/builddir/)
+target_link_directories(inventory PUBLIC ${SRC_FOLDER}/external/libdb/build_unix/)
+target_link_directories(inventory PUBLIC ${SRC_FOLDER}/external/procps/)
+
+
```
Comment (TomasTurina) [inline src/modules/inventory/CMakeLists.txt:39]: Remove extra empty line.

File: src/modules/inventory/CMakeLists.txt
Comment (TomasTurina) [inline src/modules/inventory/CMakeLists.txt:1]: Check also the usage of `set_common_settings` and `configure_target`.

File: src/modules/inventory/src/inventory.cpp [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    Inventory::instance().destroy();
+}
+
+string Inventory::command(const string & query) {
+    log(LOG_INFO, "Query: " + query);
+    return "OK";
+}
+
+string Inventory::name() const {
+    return "inventory";
```
Comment (TomasTurina) [inline src/modules/inventory/src/inventory.cpp:89]: Can this be a const string?

File: src/modules/inventory/src/inventory.cpp [hunk 8]
```diff
@@ -0,0 +1,155 @@
+/*
```
Comment (TomasTurina) [inline src/modules/inventory/src/inventory.cpp:1]: I think we can leave this comment for another PR as many files still don't have it.

File: src/modules/inventory/tests/invNormalizer/CMakeLists.txt
Comment (TomasTurina) [inline src/modules/inventory/tests/invNormalizer/CMakeLists.txt:1]: Check also the usage of `configure_target` in the tests.

File: src/modules/inventory/tests/inventoryImp/CMakeLists.txt
Comment (TomasTurina) [inline src/modules/inventory/tests/inventoryImp/CMakeLists.txt:1]: Check also the usage of `configure_target` in the tests.

File: src/modules/tests/CMakeLists.txt
Comment (TomasTurina) [inline src/modules/tests/CMakeLists.txt:1]: Check also the usage of `configure_target` in the tests.

(check for example `src/agent/tests/CMakeLists.txt`)

File: src/modules/tests/moduleManager_test.cpp
Comment (TomasTurina) [inline src/modules/tests/moduleManager_test.cpp:1]: Make sure all tests run on GHA workflows in all OS.

**Review by TomasTurina — CHANGES_REQUESTED** [review]
Are the UT of the new modules/components running in the GHA workflows (Linux, Windows and macOS)?

File: src/Makefile
Comment (TomasTurina) [inline src/Makefile:1]: Review this file, we should take care only for linux, macOS and windows (tier1).

File: src/agent/CMakeLists.txt [hunk 9]
```diff
@@ -34,7 +34,7 @@ set(SOURCES
 
 add_library(Agent ${SOURCES})
 target_include_directories(Agent PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/src)
-target_link_libraries(Agent PUBLIC ConfigurationParser Communicator AgentInfo CommandHandler MultiTypeQueue PRIVATE OpenSSL::SSL OpenSSL::Crypto Boost::asio Boost::beast Logger)
+target_link_libraries(Agent PUBLIC ConfigurationParser Communicator AgentInfo CommandHandler MultiTypeQueue ModuleManager PRIVATE OpenSSL::SSL OpenSSL::Crypto Boost::asio Boost::beast)
```
Comment (TomasTurina) [inline src/agent/CMakeLists.txt:37]: Why did you remove the logger?

File: src/Makefile
Comment (TomasTurina) [inline src/Makefile:1]: After the ports overlay, I think this is no longer necessary.

File: src/agent/configuration_parser/src/configuration_parser.cpp [hunk 10]
```diff
@@ -19,10 +19,25 @@ namespace configuration
...
-                R"([agent]
+                R"(
+                [agent]
                 server_mgmt_api_port = "55000"
                 agent_comms_api_port = "27000"
-                manager_ip = "localhost")",
+                manager_ip = "localhost"
+
+                [inventory]
+                disabled = false
```
Comment (TomasTurina) [inline src/agent/configuration_parser/src/configuration_parser.cpp:29]: Maybe it's better to name this setting `enabled`.

File: src/common/binaries_op/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,5 @@
+add_library(binaries_op STATIC src/binaries_op.c)
+
+target_include_directories(binaries_op PUBLIC include)
+
+target_link_libraries(binaries_op utils debug_op)
```
Comment (TomasTurina) [inline src/common/binaries_op/CMakeLists.txt:5]: Remember to review and fix all the missing end of file new lines.

File: src/common/data_provider/src/sharedDefs.h [hunk 12]
```diff
@@ -50,7 +46,6 @@ enum OSPlatformType
     LINUX,
     BSDBASED,
```
Comment (TomasTurina) [inline src/common/data_provider/src/sharedDefs.h:47]: Also remove this line.

File: src/common/debug_op/CMakeLists.txt
Comment (TomasTurina) [inline src/common/debug_op/CMakeLists.txt:1]: In the future we need to replace this library with the new logger.

File: src/common/file_op/src/file_op.c [hunk 13]
```diff
@@ -2202,7 +2202,9 @@ int cldir_ex_ignore(const char * name, const char ** ignore) {
 
     while (dirent = readdir(dir), dirent) {
         // Skip "." and ".."
-        if ((dirent->d_name[0] == '.' && (dirent->d_name[1] == '\0' || (dirent->d_name[1] == '.' && dirent->d_name[2] == '\0'))) || w_str_in_array(dirent->d_name, ignore)) {
+        // TODO: replace function w_str_in_array
```
Comment (TomasTurina) [inline src/common/file_op/src/file_op.c:2205]: Why is this necessary?

File: src/common/file_op/src/file_op.c [hunk 14]
```diff
@@ -3449,25 +3451,30 @@ char *w_homedir(char *arg) {
 #ifdef __MACH__
     pid_t pid = getpid();
     if (proc_pidpath(pid, buff, PATH_MAX) > 0) {
-        buff = w_strtok_r_str_delim(delim, &buff);
+        // TODO: replace function w_strtok_r_str_delim
```
Comment (TomasTurina) [inline src/common/file_op/src/file_op.c:3454]: Same question.

File: src/common/utils/include/shared.h
Comment (TomasTurina) [inline src/common/utils/include/shared.h:1]: Remove reference of `FreeBSD` in this file.

File: src/common/expression/CMakeLists.txt
Comment (TomasTurina) [inline src/common/expression/CMakeLists.txt:1]: Remove this library if it won't be used.

File: src/common/stringHelper/include/stringHelper.h
Comment (TomasTurina) [inline src/common/stringHelper/include/stringHelper.h:1]: Complete the `CMakeLists.txt` of this library.

File: src/common/string_op/CMakeLists.txt
Comment (TomasTurina) [inline src/common/string_op/CMakeLists.txt:1]: Remove this library if it won't be used.

File: src/common/validate_op/CMakeLists.txt
Comment (TomasTurina) [inline src/common/validate_op/CMakeLists.txt:1]: Remove this library if it won't be used.

**Comment by jr0me** [comment]
A few dependencies were copied into this branch in some commits and removed later on. The problem with this approach (aside that it adds a number of commits and makes reviewing a bit more difficult 😅 ) is that it increases the size of the repository unnecessarily as all of these files are now part of its git history. If possible I would recommend to remove the commits that introduced http-request, librpm, procps, etc.

File: src/modules/inventory/tests/invNormalizer/invNormalizer_test.h
Comment (TomasTurina) [inline src/modules/inventory/tests/invNormalizer/invNormalizer_test.h:1]: Rename to `.hpp`.

File: src/common/debug_op/CMakeLists.txt
[the remark this answers is not in the document]
Comment (TomasTurina) [reply src/common/debug_op/CMakeLists.txt:1]: Or maybe we should remove this library so it is not used anymore.

File: src/modules/moduleWrapper.hpp
Comment (jr0me) [inline src/modules/moduleWrapper.hpp:1]: This file is an almost exact duplicate of `src/include/moduleWrapper.hpp` but doesn't seem to be used. Is this expected?

File: src/common/data_provider/src/network/networkBSDWrapper.h
Comment (TomasTurina) [inline src/common/data_provider/src/network/networkBSDWrapper.h:1]: Is this file necessary?

File: src/common/debug_op/CMakeLists.txt
Comment (cborla) [reply src/common/debug_op/CMakeLists.txt:1]: We are working on it.

File: src/modules/moduleWrapper.hpp [hunk 15]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <functional>
+#include <string>
+#include "configuration.hpp"
+
+using namespace std;
```
Comment (TomasTurina) [inline src/modules/moduleWrapper.hpp:7]: Remove. This is not safe.

File: src/modules/include/moduleManager.hpp [hunk 16]
```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include <map>
+#include <memory>
+#include <string>
+#include <thread>
+#include <multitype_queue.hpp>
+#include <moduleWrapper.hpp>
+
+using namespace std;
```
Comment (TomasTurina) [inline src/modules/include/moduleManager.hpp:10]: Remove. This is not safe.

File: src/modules/include/moduleWrapper.hpp [hunk 17]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <functional>
+#include <string>
+#include <configuration_parser.hpp>
+
+using namespace std;
```
Comment (TomasTurina) [inline src/modules/include/moduleWrapper.hpp:7]: Remove. This is not safe.

File: src/modules/inventory/src/inventory.cpp [hunk 18]
```diff
@@ -0,0 +1,136 @@
+#include <iostream>
+#include <cjson/cJSON.h>
+#include <shared.h>
+#include <defs.h>
+#include <logging_helper.h>
+
+#include <inventory.hpp>
+#include <sysInfo.hpp>
+
+using namespace std;
```
Comment (TomasTurina) [inline src/modules/inventory/src/inventory.cpp:10]: Remove. This is not safe.

File: src/agent/tests/CMakeLists.txt [hunk 19]
```diff
@@ -4,7 +4,9 @@ add_executable(agent_test agent_test.cpp)
 configure_target(agent_test)
 target_include_directories(agent_test PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/../include ${CMAKE_CURRENT_SOURCE_DIR}/../src)
 target_link_libraries(agent_test PRIVATE Agent GTest::gtest GTest::gmock)
+if(NOT WIN32)
```
Comment (TomasTurina) [inline src/agent/tests/CMakeLists.txt:7]: Why is this?

File: src/common/data_provider/CMakeLists.txt
Comment (TomasTurina) [inline src/common/data_provider/CMakeLists.txt:1]: Replace `UNIT_TEST` with `BUILD_TESTS` so tests can run.

File: src/common/dbsync/CMakeLists.txt
Comment (TomasTurina) [inline src/common/dbsync/CMakeLists.txt:1]: Replace `UNIT_TEST` with `BUILD_TESTS` so tests can run.

File: src/common/networkHelper/CMakeLists.txt [hunk 20]
```diff
@@ -1,5 +1,5 @@
-cmake_minimum_required(VERSION 3.12.4)
+add_library(networkHelper INTERFACE)
 
-if(UNIT_TEST)
-    add_subdirectory(tests)
-endif()
+target_include_directories(networkHelper INTERFACE include)
+
+target_link_libraries(networkHelper INTERFACE utils)
```
Comment (TomasTurina) [inline src/common/networkHelper/CMakeLists.txt:5]: Missing new line.

File: src/common/logger/src/logger.c
Comment (TomasTurina) [inline src/common/logger/src/logger.c:1]: Rename to `logger.cpp`.

File: src/common/privsep_op/src/privsep_op.c [hunk 21]
```diff
@@ -178,7 +158,7 @@ int Privsep_Chroot(const char *path)
         return (OS_INVALID);
     }
 
-    nowChroot();
+    // nowChroot();
```
Comment (TomasTurina) [inline src/common/privsep_op/src/privsep_op.c:161]: Why did you commented this?

File: src/Makefile
Comment (nbertoldo) [reply src/Makefile:1]: Fixed in #186

File: src/common/logger/src/logger.c
Comment (LucioDonda) [reply src/common/logger/src/logger.c:1]: renamed

File: src/common/networkHelper/CMakeLists.txt (hunk 20 above)
Comment (LucioDonda) [reply src/common/networkHelper/CMakeLists.txt]: done

File: src/modules/inventory/src/inventory.cpp (hunk 18 above)
Comment (LucioDonda) [reply src/modules/inventory/src/inventory.cpp]: done!

File: src/modules/include/moduleWrapper.hpp (hunk 17 above)
Comment (LucioDonda) [reply src/modules/include/moduleWrapper.hpp]: done!

File: src/modules/include/moduleManager.hpp (hunk 16 above)
Comment (LucioDonda) [reply src/modules/include/moduleManager.hpp]: done!

File: src/agent/configuration_parser/src/configuration_parser.cpp (hunk 10 above)
Comment (LucioDonda) [reply src/agent/configuration_parser/src/configuration_parser.cpp]: done alongside tests and several functions.

File: src/common/string_op/CMakeLists.txt
Comment (LucioDonda) [reply src/common/string_op/CMakeLists.txt:1]: done!

File: src/modules/inventory/tests/invNormalizer/invNormalizer_test.h
Comment (LucioDonda) [reply src/modules/inventory/tests/invNormalizer/invNormalizer_test.h:1]: done!

File: src/modules/moduleWrapper.hpp
Comment (LucioDonda) [reply src/modules/moduleWrapper.hpp:1]: removed.

File: src/common/data_provider/src/sharedDefs.h (hunk 12 above)
Comment (cborla) [reply src/common/data_provider/src/sharedDefs.h:47]: This may impact on macOS compilation, we can discuss this later.

File: src/agent/tests/CMakeLists.txt (hunk 19 above)
Comment (cborla) [reply src/agent/tests/CMakeLists.txt:7]: At this stage the module for windows is not functional so we decided to discard the tests, which will be included in the issue [Support Compilation for Inventory Module on Windows Using MSVC](https://github.com/wazuh/wazuh-agent/issues/167#top)

File: src/common/data_provider/src/network/networkBSDWrapper.h
Comment (LucioDonda) [reply src/common/data_provider/src/network/networkBSDWrapper.h:1]: this file is included in `wazuh-agent/src/common/data_provider/src/network/networkInterfaceBSD.cpp`
that must be used during macos compilation as showed here:

```cmake
elseif(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
  if(${CMAKE_HOST_SYSTEM_PROCESSOR} MATCHES "arm64.*|ARM64.*")
    file(GLOB SYSINFO_SRC
        "${CMAKE_CURRENT_SOURCE_DIR}/src/*Mac.cpp"
        "${CMAKE_CURRENT_SOURCE_DIR}/src/*CommonBSD.cpp"
        "${CMAKE_CURRENT_SOURCE_DIR}/src/packages/*Mac.cpp"
        "${CMAKE_CURRENT_SOURCE_DIR}/src/network/*BSD.cpp"
        "${CMAKE_CURRENT_SOURCE_DIR}/src/osinfo/sysOsParsers.cpp"
        "${CMAKE_CURRENT_SOURCE_DIR}/src/hardware/*ARMMac.cpp")
  else()
```

**Review by TomasTurina — APPROVED** [review]

**Review by cborla — APPROVED** [review]
LGTM!
